### PR TITLE
Feat/ecs/hitreg

### PR DESCRIPTION
**Pull request by aurelenc**
Add hit registration systems
- Hitbox detection
- Faction system
- Add projectile speed to weapons
- Fix movement bug in non-controllable entities

### CODE REVIEW

File: src/ecs/Registry.hpp [hunk 1]
```diff
@@ -103,6 +105,8 @@ namespace ecs
          */
         void killEntity(Entity const &e)
         {
+            if (e._id == ecs::constant::npos)
+                return;
             for (const auto &erase_func : _eraseFunctions)
                 erase_func(*this, e);
```
Comment (Cavonstavant) [inline src/ecs/Registry.hpp:111]:
```suggestion
                eraseFunc(*this, e);
```

**Review by Cavonstavant — CHANGES_REQUESTED** [review]

**Review by Cavonstavant — APPROVED** [review]

File: src/ecs/systems/ProjectileCollision.hpp [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+#pragma endregion
+
+#pragma region Hitreg
+        auto &healths = world.registry.getComponents<component::Health>();
+        auto const &factions = world.registry.getComponents<component::Faction>();
+        auto const &sizes = world.registry.getComponents<component::Size>();
+        bool killProj = false;
+        size_t projId = 0;
+
+        for (size_t i = 0; i < projectiles.size() && i < positions.size(); ++i) {
```
Comment (Bootoyka) [inline src/ecs/systems/ProjectileCollision.hpp:37]: 1

File: src/ecs/systems/ProjectileCollision.hpp [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+            auto const &proj = projectiles[i];
+            auto const &projPos = positions[i];
+
+            if (proj && projPos) {
```
Comment (Bootoyka) [inline src/ecs/systems/ProjectileCollision.hpp:41]: 2

File: src/ecs/systems/ProjectileCollision.hpp [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+                for (size_t j = 0; j < healths.size() && j < positions.size(); ++j) {
```
Comment (Bootoyka) [inline src/ecs/systems/ProjectileCollision.hpp:42]: 3

File: src/ecs/systems/ProjectileCollision.hpp [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+                    auto &victimHealth = healths[j];
+                    auto const &victimPos = positions[j];
+                    auto const &victimSize = sizes[j];
+
+                    if (victimHealth && victimPos && victimSize) {
```
Comment (Bootoyka) [inline src/ecs/systems/ProjectileCollision.hpp:47]: 4

File: src/ecs/systems/ProjectileCollision.hpp [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+                        if (i < factions.size() && j < factions.size())
```
Comment (Bootoyka) [inline src/ecs/systems/ProjectileCollision.hpp:48]: 5

File: src/ecs/systems/ProjectileCollision.hpp [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+                            if (factions[i] && factions[j])
```
Comment (Bootoyka) [inline src/ecs/systems/ProjectileCollision.hpp:49]: 6

File: src/ecs/systems/ProjectileCollision.hpp [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+                                if (factions[i].value().faction == factions[j].value().faction)
```
Comment (Bootoyka) [inline src/ecs/systems/ProjectileCollision.hpp:50]: 7 conditional branching

**Review by Bootoyka — CHANGES_REQUESTED** [review]
i can't accept 7 conditional branch, for the rest, lgtm i need to test by my side

**Comment by aurelenc** [comment]
We game Dev or we not game Dev?

File: src/ecs/systems/ProjectileCollision.hpp (hunk 8 above)
Comment (aurelenc) [reply src/ecs/systems/ProjectileCollision.hpp:50]: wontfix

**Pull request merged**